### PR TITLE
Re-iteration on PR #5134; multiple optimizations from liberty

### PR DESCRIPTION
**Pull request by Yatiyaya**
## About The Pull Request
<details>
<summary>
	A general refurbish of the PR #5134. Updated with planned ideas that had yet to be implimented. 
Currently, in the test phase, as mob Subsystem has been designed to never sleep due to tickrate (HIGHLY FUCKING EXPERIMENTAL)

Using this as a platform PR to initiate the start of rolling lag fixes.
</summary>
<hr>

	
<hr>
</details>

## Changelog
:cl:

tweak: tweaked a a lot of things....

/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/slime/life.dm [hunk 1]
```diff
@@ -348,10 +348,10 @@
...
 			UnarmedAttack(frenemy)
 
-	var/sleeptime = movement_delay()
-	if(sleeptime <= 5) sleeptime = 5 // Maximum one action per half a second
-	spawn (sleeptime)
-		handle_AI()
+//	var/sleeptime = movement_delay()
+//	if(sleeptime <= 5) sleeptime = 5 // Maximum one action per half a second
+//	spawn (sleeptime)
+//		handle_AI()
```
Comment (Yatiyaya) [inline code/modules/mob/living/carbon/slime/life.dm:354]: Recursive spawn call on itself, when it's already procced by life()....
WHY?!

File: code/modules/mob/living/carbon/superior_animal/ai.dm [hunk 2]
```diff
@@ -30,18 +22,16 @@
 
 	var/turf/our_turf = get_turf(src)
 	if (our_turf) //If we're not in anything, continue
-		for(var/mob/living/target_mob as anything in hearers(src, viewRange)) //as anything works because isvalid has a isliving check, change if necessary
-			if (isValidAttackTarget(target_mob))
+		for(var/mob/living/target_mob as anything in hearers(viewRange, src)) //as anything works because isvalid has a isliving check, change if necessary
+			if(!(target_mob?.faction == faction && !attack_same) && isValidAttackTarget(target_mob))
```
Comment (Yatiyaya) [inline code/modules/mob/living/carbon/superior_animal/ai.dm:26]: Sanity check.
Don't bother checking for valid. Shaves off a single proc for hordes.
Could be done better, if the for() is rewritten.

File: code/modules/mob/living/carbon/superior_animal/ameridian/ameridian_golem.dm [hunk 3]
```diff
@@ -77,11 +77,12 @@
 			var/obj/item/projectile/sonic_bolt/SB = P
 			SB.multiply_projectile_damage(SB.golem_damage_bonus)
 			drop_amount = 0 // No loot
+	spawn(100 MILLISECONDS)
+		if(!QDELETED(src))
+			maintain_drop_amount()
```
Comment (Yatiyaya) [inline code/modules/mob/living/carbon/superior_animal/ameridian/ameridian_golem.dm:82]: addtimer(callback() was being pinged by crash()

File: code/modules/mob/living/carbon/superior_animal/superior_animal.dm [hunk 4]
```diff
@@ -429,7 +409,7 @@
 			patience = patience_initial
 		// This block controls our attack/range logic
 		var/atom/targetted = targetted_mob
-		if (!(targetted_mob.check_if_alive(TRUE)))
+		if (!(targetted_mob?.check_if_alive(TRUE)))
```
Comment (Yatiyaya) [inline code/modules/mob/living/carbon/superior_animal/superior_animal.dm:412]: becaused targetted can be null.

File: code/modules/mob/living/carbon/superior_animal/superior_animal.dm [hunk 5]
```diff
@@ -575,66 +533,18 @@
...
-
-	if(life_cycles_before_scan)
-		life_cycles_before_scan--
-		return FALSE
-	if(check_surrounding_area(viewRange))
-		activate_ai()
-		life_cycles_before_scan = initial(life_cycles_before_scan)/6 //So it doesn't fall asleep just to wake up the next tick
-		return TRUE
-	life_cycles_before_scan = initial(life_cycles_before_scan)
-	return FALSE
```
Comment (Yatiyaya) [inline code/modules/mob/living/carbon/superior_animal/superior_animal.dm:625]: Was already being called on another Life() parrent. No need to repeat this, since the whole function is now moved up.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 6]
```diff
@@ -217,9 +217,10 @@ var/list/mydirs = list(NORTH, SOUTH, EAST, WEST, SOUTHWEST, NORTHWEST, NORTHEAST
 /mob/living/simple_animal/hostile/proc/ListTargets(var/dist = 7)
 	var/list/L = hearers(src, dist)
 
-	for (var/obj/mecha/M in GLOB.mechas_list)
-		if (M.z == src.z && get_dist(src, M) <= dist)
-			L += M
+	for(var/atom/movable/AM in view(dist, src))
+		var/M = AM.get_mob()	// we look inside things
+		if(M)
+			LAZYDISTINCTADD(L, M)
```
Comment (Yatiyaya) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm:223]: get_mob() already handles mecha occupant checks.
GLOB.mechas_list checks every mecha, everywhere, every Zlevel...

Stinky

File: code/modules/sanity/sanity_mob.dm [hunk 7]
```diff
@@ -44,7 +44,8 @@ GLOBAL_VAR_INIT(GLOBAL_INSIGHT_MOD, 1)
 
 
 #define EAT_COOLDOWN_MESSAGE 15 SECONDS
-#define SANITY_MOB_DISTANCE_ACTIVATION 12
+//used to be 12
+#define SANITY_MOB_DISTANCE_ACTIVATION world.view
```
Comment (Yatiyaya) [inline code/modules/sanity/sanity_mob.dm:48]: Honestly, removed the proc() that uses this.

But, world.view can be expressed X or "XxY"

**Comment by Yatiyaya** [comment]
weird... Runtiming on a callback that's not on my test server.

**Comment by cdb-is-not-good** [comment]
A number of issues and whatnot were traced back to this PR and I'm led to believe the creator was well aware but didn't bother to communicate such. If you need profiler data from a PR you know is going to cause issues we need to be made aware of that in the future if you want your PRs to be considered.

**Comment by Yatiyaya** [comment]
Don't make false accusations to enlarge your own fucking ego, you cocksucker.

**Comment by Yatiyaya** [comment]
This is why you are the fucking joke of sojourn, along with tribly. The moment something happens, you try to undermine other people to boost your own god damn self-esteem.

Go find a ditch to fill, because you obviously have no other use in life, so find one while being dead.

**Comment by Pawoverlord** [comment]
> This is why you are the fucking joke of sojourn, along with tribly. The moment something happens, you try to undermine other people to boost your own god damn self-esteem.
> 
> Go find a ditch to fill, because you obviously have no other use in life, so find one while being dead.
> 
> The fact you even had to post a comment like that, just affirms you have no god damn idea how coding works, in the slightest.

Right-- firstly. Don't make comments like this. I get it and understand you're upset, but wishing death upon someone isn't okay. Don't do it. It _very quickly_ loses my favor on these topics.

Additionally, doing testing on personal versions of the server can help aid in the testing process if you don't already. I'm not sure of what's going on with this merge or have any understanding of why CDB is shotgunning at you.


> A number of issues and whatnot were traced back to this PR and I'm led to believe the creator was well aware but didn't bother to communicate such. If you need profiler data from a PR you know is going to cause issues we need to be made aware of that in the future if you want your PRs to be considered.

As for you, CDB. Please provide the issues here so that they can be corrected if they are actually issues. It does no one _any_ good to not discuss the issues unless you wanted to flat out reject the PR. Which I'm not going to object too harshly on, but at least provide some input other than not bothering to communicate. Testing does come with testing. But, you are a head of the coding staff; use this as a potential teaching moment to allow our coders that are willing to take _time out of their day_ to come help us with our project. They do deserve to know what's causing difficulties that you're finding. 

TL:DR - Please. Try to be understanding, we're not all professionals. We're going to make mistakes with garbage code.

File: code/modules/mob/living/carbon/superior_animal/ai.dm (hunk 2 above)
Comment (king5327) [reply code/modules/mob/living/carbon/superior_animal/ai.dm:26]: Most of the overhead from calling a proc is in the actual running of that proc. If that proc changes we want all the places that use that logic to update with it. This isn't so much an optimization as it is a gutting.
